boa/abi/function.py

```diff
@@ -0,0 +1,160 @@
+from functools import cached_property
+from typing import TYPE_CHECKING, Optional
+
+from eth_abi import decode, encode, is_encodable
+from vyper.semantics.analysis.base import FunctionVisibility, StateMutability
+from vyper.utils import method_id
+
+if TYPE_CHECKING:
+    from boa.abi.contract import ABIContract
+
+
+class _EvmFunction:
+    ...
+
+
+class ABIFunction(_EvmFunction):
+    def __init__(self, abi: dict, contract_name: str):
+        super().__init__()
+        self._abi = abi
+        self._contract_name = contract_name
+        self._function_visibility = FunctionVisibility.EXTERNAL
+        self._mutability = StateMutability.from_abi(abi)
+        self.contract: Optional["ABIContract"] = None
+
+    @property
+    def name(self):
+        return self._abi["name"]
+
+    @cached_property
+    def argument_types(self) -> list[str]:
+        return [i["type"] for i in self._abi["inputs"]]
+
+    @property
+    def argument_count(self) -> int:
+        return len(self.argument_types)
+
+    @property
+    def signature(self) -> str:
+        return f"({','.join(self.argument_types)})"
+
+    @cached_property
+    def return_type(self) -> list[str]:
+        return [o["type"] for o in self._abi["outputs"]]
+
+    @property
+    def full_signature(self) -> str:
+        return f"{self.name}{self.signature} -> {self.return_type}"
+
+    @cached_property
+    def method_id(self) -> bytes:
+        return method_id(self.name + self.signature)
+
+    def __repr__(self) -> str:
+        return f"ABI {self._contract_name}.{self.full_signature}"
+
+    def __str__(self) -> str:
+        return repr(self)
+
+    @property
+    def is_mutable(self) -> bool:
+        return self._mutability > StateMutability.VIEW
+
+    def __eq__(self, other):
+        return isinstance(other, ABIFunction) and self._abi == other._abi
+
+    def __hash__(self):
+        return hash(self._abi)
+
+    def is_encodable(self, *args, **kwargs) -> bool:
+        """Check whether this function accepts the given arguments after eventual encoding."""
+        parsed_args = self._merge_kwargs(*args, **kwargs)
+        return all(
+            is_encodable(abi_type, arg)
+            for abi_type, arg in zip(self.argument_types, parsed_args)
```

Comment:
`_merge_kwargs()` checks if `len(self.argument_types) == len(parsed_args)` right? i think there could be an assert here anyways, as the use of `zip()` without sanity checking argument lengths makes me nervous

Reply:
https://github.com/vyperlang/titanoboa/blob/71058ad9c430322b9de7e838f82d1dd81d3a20dc/boa/abi.py#L68